presto-main/src/main/java/com/facebook/presto/operator/scalar/ArrayJoin.java

```diff
@@ -0,0 +1,149 @@
+    }
+
+    @Override
+    public boolean isHidden()
+    {
+        return false;
+    }
+
+    @Override
+    public boolean isDeterministic()
+    {
+        return true;
+    }
+
+    @Override
+    public String getDescription()
+    {
+        return "Concatenates the elements of the given array using a delimiter and an optional string to replace nulls";
+    }
+
+    @Override
+    public FunctionInfo specialize(Map<String, Type> types, int arity, TypeManager typeManager, FunctionRegistry functionRegistry)
+    {
+        Type type = types.get("T");
+        TypeSignature valueType = type.getTypeSignature();
+        TypeSignature arrayType = parameterizedTypeName(StandardTypes.ARRAY, valueType);
+        Signature signature = new Signature(FUNCTION_NAME, VARCHAR_TYPE_SIGNATURE, arrayType, VARCHAR_TYPE_SIGNATURE, VARCHAR_TYPE_SIGNATURE);
+
+        MethodHandle methodHandle = methodHandle(ArrayJoin.class, "arrayJoin", FunctionInfo.class, Type.class, ConnectorSession.class, Slice.class, Slice.class, Slice.class);
+        FunctionInfo castFunction = functionRegistry.getExactFunction(internalOperator(CAST.name(), VARCHAR_TYPE_SIGNATURE, ImmutableList.of(valueType)));
+
+        return new FunctionInfo(signature, getDescription(), isHidden(), methodHandle.bindTo(castFunction).bindTo(type), isDeterministic(), false, ImmutableList.of(false, false, true));
+    }
+
+    public static Slice arrayJoin(FunctionInfo castFunction, Type elementType, ConnectorSession session, Slice array, Slice delimiter, Slice nullReplacement) throws Throwable
+    {
+        if (elementType instanceof ArrayType || elementType instanceof MapType || elementType instanceof RowType) {
+            throw new PrestoException(INVALID_FUNCTION_ARGUMENT, format("Unexpected element type %s", elementType.getDisplayName()));
+        }
+
+        Block arrayBlock = readStructuralBlock(array);
+        int numElements = arrayBlock.getPositionCount();
+        DynamicSliceOutput sliceOutput = new DynamicSliceOutput(array.length());
+        Class<?> javaType = elementType.getJavaType();
+
+        for (int i = 0; i < numElements; i++) {
+            if (arrayBlock.isNull(i)) {
+                if (nullReplacement != null) {
+                    sliceOutput.appendBytes(nullReplacement);
+                }
+            }
+            else {
+                if (javaType == boolean.class) {
+                    sliceOutput.appendBytes((Slice) castFunction.getMethodHandle().invokeWithArguments(elementType.getBoolean(arrayBlock, i)));
+                }
+                else if (javaType == double.class) {
+                    sliceOutput.appendBytes((Slice) castFunction.getMethodHandle().invokeWithArguments(elementType.getDouble(arrayBlock, i)));
+                }
+                else if (javaType == long.class) {
+                    long value = elementType.getLong(arrayBlock, i);
+                    Slice slice;
+                    if (elementType == TIMESTAMP) {
```

Comment:
Instead of a check for `TIMESTAMP` check MethodHandle.type().parameterArray() and invoke it with the session, if necessary, for all these if branches